rust/automerge/src/transaction/inner.rs

```diff
@@ -447,30 +521,142 @@ impl TransactionInner {
     pub(crate) fn splice<Obs: OpObserver>(
         &mut self,
         doc: &mut Automerge,
-        mut op_observer: Option<&mut Obs>,
+        op_observer: Option<&mut Obs>,
         ex_obj: &ExId,
-        mut pos: usize,
+        pos: usize,
         del: usize,
         vals: impl IntoIterator<Item = ScalarValue>,
     ) -> Result<(), AutomergeError> {
         let obj = doc.exid_to_obj(ex_obj)?;
-        for _ in 0..del {
-            // This unwrap and rewrap of the option is necessary to appeas the borrow checker :(
-            if let Some(obs) = op_observer.as_mut() {
-                self.local_op(doc, Some(*obs), obj, pos.into(), OpType::Delete)?;
+        let obj_type = doc
+            .ops
+            .object_type(&obj)
+            .ok_or(AutomergeError::NotAnObject)?;
+        if obj_type != ObjType::List {
+            return Err(AutomergeError::InvalidOp(obj_type));
+        }
+        let vals = vals.into_iter().collect();
+        self.inner_splice(doc, op_observer, obj, pos, del, vals, SpliceType::List)
+    }
+
+    /// Splice string into a text object
+    pub(crate) fn splice_text<Obs: OpObserver>(
+        &mut self,
+        doc: &mut Automerge,
+        op_observer: Option<&mut Obs>,
+        ex_obj: &ExId,
+        pos: usize,
+        del: usize,
+        text: &str,
+    ) -> Result<(), AutomergeError> {
+        let obj = doc.exid_to_obj(ex_obj)?;
+        let obj_type = doc
+            .ops
+            .object_type(&obj)
+            .ok_or(AutomergeError::NotAnObject)?;
+        if obj_type != ObjType::Text {
+            return Err(AutomergeError::InvalidOp(obj_type));
+        }
+        let vals = text.chars().map(ScalarValue::from).collect();
+        self.inner_splice(
+            doc,
+            op_observer,
+            obj,
+            pos,
+            del,
+            vals,
+            SpliceType::Text(text, doc.utf16),
+        )
+    }
+
+    #[allow(clippy::too_many_arguments)]
```

Comment:
I think it would be worth adding a `SpliceArgs` struct which holds all these arguments. We call this function in three places, adding a struct will make refactoring those calls easier.